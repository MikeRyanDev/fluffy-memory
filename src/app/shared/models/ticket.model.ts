import { UserModel } from './user.model';

export interface TicketModel {
  id: number;
  description: string;
  assigneeId: number;
  completed: boolean;
}

export interface TicketViewModel extends TicketModel {
  assignee: UserModel;
}
