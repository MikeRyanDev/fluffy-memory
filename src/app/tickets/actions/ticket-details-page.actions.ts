import { Action } from '@ngrx/store';

export enum Types {
  Enter = '[Ticket Details Page Enter',
}

export interface EnterAction extends Action {
  type: Types.Enter;
}

export function createEnterAction(): EnterAction {
  return { type: Types.Enter };
}

export type Union = EnterAction;
