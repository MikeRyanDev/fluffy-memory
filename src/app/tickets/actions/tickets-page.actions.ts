import { Action } from '@ngrx/store';

export enum Types {
  Enter = '[Tickets Page] Enter',
  ToggleTicketComplete = '[Tickets Page] Toggle Ticket Complete',
  FilterByDescription = '[Tickets Page] Filter by Description',
  FilterByUser = '[Tickets Page] Filter by User',
}

export interface EnterAction extends Action {
  type: Types.Enter;
}

export function createEnterAction(): EnterAction {
  return { type: Types.Enter };
}

export interface ToggleTicketCompleteAction extends Action {
  type: Types.ToggleTicketComplete;
  ticketId: number;
  isComplete: boolean;
}

export function createToggleTicketCompleteAction(
  ticketId: number,
  isComplete: boolean,
): ToggleTicketCompleteAction {
  return { type: Types.ToggleTicketComplete, ticketId, isComplete };
}

export interface FilterByDescriptionAction {
  type: Types.FilterByDescription;
  searchTerm: string;
}

export function createFilterByDescriptionAction(
  searchTerm: string,
): FilterByDescriptionAction {
  return { type: Types.FilterByDescription, searchTerm };
}

export interface FilterByUserAction {
  type: Types.FilterByUser;
  userId: number;
}

export function createFilterByUserAction(userId: number): FilterByUserAction {
  return { type: Types.FilterByUser, userId };
}

export type Union =
  | EnterAction
  | ToggleTicketCompleteAction
  | FilterByDescriptionAction
  | FilterByUserAction;
