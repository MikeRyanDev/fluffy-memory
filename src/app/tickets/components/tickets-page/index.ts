export * from './tickets-page.component';
export * from './tickets-page.module';
