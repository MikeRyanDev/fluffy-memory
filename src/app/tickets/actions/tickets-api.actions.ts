import { Action } from '@ngrx/store';
import { UserModel, TicketModel } from '../../shared/models';

export enum Types {
  LoadUsersAndTicketsSuccess = '[Tickets API] Load Users and Tickets Success',
  LoadUsersAndTicketsFailure = '[Tickets API] Load Users and Tickets Failure',
  SaveCompleteStatusSuccess = '[Tickets API] Save Complete Status Success',
  SaveCompleteStatusFailure = '[Tickets API] Save Complete Status Failure',
  RetrySaveCompleteStatus = '[Tickets API] Retry Save Complete Status',
}

export interface LoadUsersAndTicketsSuccessAction extends Action {
  type: Types.LoadUsersAndTicketsSuccess;
  users: UserModel[];
  tickets: TicketModel[];
}

export function createLoadUsersAndTicketsSuccessAction(
  users: UserModel[],
  tickets: TicketModel[],
): LoadUsersAndTicketsSuccessAction {
  return { type: Types.LoadUsersAndTicketsSuccess, users, tickets };
}

export interface LoadUsersAndTicketsFailureAction extends Action {
  type: Types.LoadUsersAndTicketsFailure;
}

export function createLoadUsersAndTicketsFailureAction(): LoadUsersAndTicketsFailureAction {
  return { type: Types.LoadUsersAndTicketsFailure };
}

export interface SaveCompleteStatusSuccessAction extends Action {
  type: Types.SaveCompleteStatusSuccess;
  ticketId: number;
}

export function createSaveCompleteStatusSuccessAction(
  ticketId: number,
): SaveCompleteStatusSuccessAction {
  return { type: Types.SaveCompleteStatusSuccess, ticketId };
}

export interface SaveCompleteStatusFailureAction extends Action {
  type: Types.SaveCompleteStatusFailure;
  ticketId: number;
  isComplete: boolean;
}

export function createSaveCompleteStatusFailureAction(
  ticketId: number,
  isComplete: boolean,
): SaveCompleteStatusFailureAction {
  return { type: Types.SaveCompleteStatusFailure, ticketId, isComplete };
}

export interface RetrySaveCompleteStatusAction {
  type: Types.RetrySaveCompleteStatus;
  ticketId: number;
  isComplete: boolean;
}

export function createRetrySaveCompleteStatusAction(
  ticketId: number,
  isComplete: boolean,
): RetrySaveCompleteStatusAction {
  return { type: Types.RetrySaveCompleteStatus, ticketId, isComplete };
}

export type Union =
  | LoadUsersAndTicketsSuccessAction
  | LoadUsersAndTicketsFailureAction
  | SaveCompleteStatusSuccessAction
  | SaveCompleteStatusFailureAction
  | RetrySaveCompleteStatusAction;
