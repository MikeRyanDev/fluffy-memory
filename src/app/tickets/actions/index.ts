import * as TicketDetailsPageActions from './ticket-details-page.actions';
import * as TicketsApiActions from './tickets-api.actions';
import * as TicketsPageActions from './tickets-page.actions';

export { TicketDetailsPageActions, TicketsApiActions, TicketsPageActions };
