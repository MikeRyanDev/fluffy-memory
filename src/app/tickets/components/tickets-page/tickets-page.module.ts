import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatProgressBarModule } from '@angular/material';
import { TicketListComponentModule } from '../ticket-list';
import { TicketsPageFiltersComponentModule } from '../tickets-page-filters';
import { TicketsPageComponent } from './tickets-page.component';

@NgModule({
  imports: [
    CommonModule,
    MatProgressBarModule,
    TicketListComponentModule,
    TicketsPageFiltersComponentModule,
  ],
  declarations: [TicketsPageComponent],
  exports: [TicketsPageComponent],
  entryComponents: [TicketsPageComponent],
})
export class TicketsPageComponentModule {}
