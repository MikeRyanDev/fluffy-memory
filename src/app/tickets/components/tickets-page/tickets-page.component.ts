import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import * as fromRoot from '../../../shared/state';
import { TicketViewModel } from '../../../shared/models';
import { TicketsPageActions } from '../../actions';

@Component({
  selector: 'app-tickets-page',
  template: `
    <app-tickets-page-filters></app-tickets-page-filters>
    <mat-progress-bar
      color="accent"
      mode="indeterminate"
      *ngIf="isLoading$ | async"
    ></mat-progress-bar>
    <app-ticket-list [tickets]="tickets$ | async" (toggleComplete)="onToggleComplete($event)">
    </app-ticket-list>
  `,
})
export class TicketsPageComponent implements OnInit {
  tickets$: Observable<TicketViewModel[]>;
  isLoading$: Observable<boolean>;

  constructor(private store: Store<fromRoot.State>) {
    this.tickets$ = store.select(fromRoot.selectFilteredTickets);
    this.isLoading$ = store.select(fromRoot.selectIsTicketsPageLoading);
  }

  ngOnInit() {
    this.store.dispatch(TicketsPageActions.createEnterAction());
  }

  onToggleComplete(ticket: TicketViewModel) {
    this.store.dispatch(
      TicketsPageActions.createToggleTicketCompleteAction(ticket.id, !ticket.completed),
    );
  }
}
