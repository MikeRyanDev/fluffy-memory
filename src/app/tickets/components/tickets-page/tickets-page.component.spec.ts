import { NO_ERRORS_SCHEMA } from '@angular/core';
import { TestBed, ComponentFixture } from '@angular/core/testing';
import { CommonModule } from '@angular/common';
import { By } from '@angular/platform-browser';
import { Store } from '@ngrx/store';
import { provideMockStore } from '@ngrx/store/testing';
import { of } from 'rxjs';
import { TicketViewModel } from 'src/app/shared/models';
import { TicketsPageActions } from '../../actions';
import { TicketsPageComponent } from './tickets-page.component';

describe('Tickets Page Component', () => {
  const mockTicket: TicketViewModel = {
    id: 3,
    description: 'Another Test Ticket',
    completed: false,
    assigneeId: 111,
    assignee: {
      id: 111,
      name: 'Mock User',
    },
  };

  let mockStore: Store<any>;

  function compile({
    tickets = [],
    isLoading = false,
  }: {
    tickets?: TicketViewModel[];
    isLoading?: boolean;
  } = {}): ComponentFixture<TicketsPageComponent> {
    const fixture = TestBed.createComponent(TicketsPageComponent);

    fixture.componentInstance.tickets$ = of(tickets);
    fixture.componentInstance.isLoading$ = of(isLoading);

    fixture.detectChanges();

    return fixture;
  }

  beforeEach(async () => {
    TestBed.configureTestingModule({
      imports: [CommonModule],
      declarations: [TicketsPageComponent],
      providers: [provideMockStore({})],
      schemas: [NO_ERRORS_SCHEMA],
    });

    mockStore = TestBed.get(Store);

    spyOn(mockStore, 'dispatch');

    await TestBed.compileComponents();
  });

  it('should render correctly', () => {
    const fixture = compile();

    expect(fixture).toMatchSnapshot();
  });

  it('should show a progress bar if the page is loading', () => {
    const fixture = compile({ isLoading: true });

    expect(fixture).toMatchSnapshot();
  });

  it('should bind the list of tickets to the ticket list component', () => {
    const mockTickets = [mockTicket];
    const fixture = compile({ tickets: mockTickets });

    const { tickets } = fixture.debugElement.query(By.css('app-ticket-list')).properties;

    expect(tickets).toEqual(mockTickets);
  });

  it('should dispatch a "toggle ticket complete" action when a ticket in the list is toggled', () => {
    const mockTickets = [mockTicket];
    const fixture = compile({ tickets: mockTickets });
    const expectedAction = TicketsPageActions.createToggleTicketCompleteAction(
      mockTicket.id,
      !mockTicket.completed,
    );

    fixture.debugElement
      .query(By.css('app-ticket-list'))
      .triggerEventHandler('toggleComplete', mockTicket);

    expect(mockStore.dispatch).toHaveBeenCalledWith(expectedAction);
  });
});
