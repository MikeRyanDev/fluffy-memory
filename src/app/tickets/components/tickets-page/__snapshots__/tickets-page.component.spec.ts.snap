// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tickets Page Component should render correctly 1`] = `
<app-tickets-page
  isLoading$={[Function Observable]}
  store={[Function MockStore]}
  tickets$={[Function Observable]}
>
  <app-tickets-page-filters /><app-ticket-list />
</app-tickets-page>
`;

exports[`Tickets Page Component should show a progress bar if the page is loading 1`] = `
<app-tickets-page
  isLoading$={[Function Observable]}
  store={[Function MockStore]}
  tickets$={[Function Observable]}
>
  <app-tickets-page-filters /><mat-progress-bar
    color="accent"
    mode="indeterminate"
  /><app-ticket-list />
</app-tickets-page>
`;
